### PR TITLE
Stabilize Arc::{increment,decrement}_strong_count

### LINKED ISSUE #71983: Tracking Issue for feature arc_mutate_strong_count
The feature gate for the issue is `#![feature(arc_mutate_strong_count)]`.

### Steps

- [x] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [x] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

- [ ] Can we improve the docs phrasing on `Arc::decr_strong_count`?

### Implementation history

- Initial PR in https://github.com/rust-lang/rust/pull/70733
- Stabilization PR in https://github.com/rust-lang/rust/pull/79285

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Tracking issue: https://github.com/rust-lang/rust/issues/71983

Stabilizes `Arc::{incr,decr}_strong_count`, enabling unsafely incrementing an decrementing the Arc strong count directly with fewer gotchas. This API was first introduced on nightly six months ago, and has not seen any changes since. The initial PR showed two existing pieces of code that would benefit from this API, and included a change inside the stdlib to use this.

Given the small surface area, predictable use, and no changes since introduction, I'd like to propose we stabilize this.

closes https://github.com/rust-lang/rust/issues/71983
r? @Mark-Simulacrum 

## Links
 * [Initial implementation](https://github.com/rust-lang/rust/pull/70733)
 * [Motivation from #68700](https://github.com/rust-lang/rust/pull/68700#discussion_r396169064)
 * [Real world example in an executor](https://docs.rs/extreme/666.666.666666/src/extreme/lib.rs.html#13)

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I am forgetting now -- is there a reason we didn't add this on Rc? I don't quickly see any discussion about that, but it seems like they'd be equally useful there ergonomics wise, though it's somewhat rarer (I suspect) to be stashing away Rc's in atomics or whatever where you're hiding them.

**Comment by yoshuawuyts** [comment]
@Mark-Simulacrum I don't believe we discussed it, and there's no reason not to as far as I recall — I'm planning to add it to Rc and both `Weak` variants as a follow-up.

If there are no objections to stabilizing the `Arc` methods though, it'd be nice if we could stabilize that sooner.

**Comment by Mark-Simulacrum** [comment]
This stabilizes the following methods on `Arc`, intended for manipulating the strong reference count without error-prone code required before (e.g., cloning the arc and forgetting it to increment or double-dropping an arc to decrement).

```rust
impl<T: ?Sized> Arc<T> {
    pub unsafe fn incr_strong_count(ptr: *const T);
    pub unsafe fn decr_strong_count(ptr: *const T);
}
```

The only reason these APIs are unsafe is that we take `*const T` rather than an `Arc<T>` or `&Arc<T>`; they are almost always used after `Arc::into_raw`, so this unsafety is deemed warranted. If we were to take an actual instance of `Arc<T>` the methods would be less useful as the caller would need to be careful to not re-drop the Arc in the case of `&Arc<T>` and/or not clone the Arc on passing it in in the owned case. Generally it's expected that this API is easier to use.

r? @dtolnay to kick off libs FCP on this.

**Comment by nagisa** [comment]
Should `incr` and `decr` be expanded to a proper word before stabilizing?

**Comment by yoshuawuyts** [comment]
@nagisa I find that `increment_strong_count` and `decrement_strong_count` feel rather verbose and would prefer we keep the naming as-is. The current names have a nice side effect that they could also stand for "increase" and "decrease" which may make them easier to memorize.

**Comment by dtolnay** [comment]
I am on board with these. #70733 has compelling discussion that this functionality is worth exposing in the standard library, and that these signatures involving `*const T` are the best ones.

@rfcbot fcp merge

**Comment by BurntSushi** [comment]
RE naming: I do have a mild preference for `increment`/`decrement` instead of abbreviating them. While these are useful, they are niche, so I don't really buy the verbosity argument. I don't think it's that important to reduce verbosity for infrequently used routines.

RE docs: Would it be possible to add some clarity to the docs explaining specifically why someone might want to use these routines?

**Comment by tmiasko** [comment]
Personally I found those functions to be rather underwhelming. They cater to relatively niche use-cases, while having a trivial implementation. Furthermore, their own implementation uses a pattern which is much more generally applicable:

* decrement: `Arc::from_raw(ptr);`
* increment: `let _ = ManuallyDrop::new(Arc::from_raw(ptr)).clone();`

**Comment by Mark-Simulacrum** [comment]
I would call neither of those trivial or easy to follow -- if I see those in the wild, I'm definitely going to expect a comment on them explaining them. I also don't think we guarantee that they will work (there's not really a way we could break that even if we wanted to and ignoring stability guarantees); I've always been rather uncertain about the "clone to increment" pattern externally from std, while internally we can of course use it.

**Comment by tmiasko** [comment]
> I also don't think we guarantee that they will work (there's not really a way we could break that even if we wanted to and ignoring stability guarantees); I've always been rather uncertain about the "clone to increment" pattern externally from std, while internally we can of course use it.

I don't think there is anything contentious about `Arc::from_raw` or `clone`. The only part that could be controversial is using `ManuallyDrop` instead of `Arc::into_raw`.

**Comment by m-ou-se** [comment]
@rfcbot concern naming

See <https://github.com/rust-lang/rust/pull/79285#issuecomment-739438411>.

I agree not abbreviating the names would be a bit better. Registering it as a concern so we don't forget.

**Comment by withoutboats** [comment]
I struggled with the pointer aspect of this API a bit. It seemed strange to combine two steps - offseting to get an arc, and then manipulating the strong count - into one function. Especially so because incrementing the strong count is safe. I was worried that these were overly-tailored to implementing a waker.

What won me over to the current API was the drop problem: users who have a raw pointer have to make sure if they get an Arc from it that they know not to drop it. This is really tricky to get right because its all implicit; implementing it properly in std seems like the right choice.

**Comment by yoshuawuyts** [comment]
It seems everyone on the libs team has ticked their box! -- but the one outstanding concern is naming. So I'd like to ask: does anyone feel strongly **against** `increment_strong_count`, `decrement_strong_count`?

I've already expressed I prefer the short versions, but I don't feel strong enough about them to block this PR on that. So I want to check if anyone objects to the long-form instead. If not, I'll happily switch this PR over to use `increment_strong_count`, `decrement_strong_count` so it can be merged. Thanks!

**Comment by yoshuawuyts** [comment]
I've update the PR to use the `{increment,decrement}_strong_count` method names. I believe once CI passes this should be ready for FCP.

**Comment by m-ou-se** [comment]
@rfcbot resolve naming

**Comment by yoshuawuyts** [comment]
Oops, I forgot to update the usage of this API in `task.rs` -- fixed and pushed again!

**Comment by CAD97** [comment]
FWIW, I use `clone_raw(*const T) -> Arc<T>` in [my real-world implementation of "+0 `Arc`"](https://www.github.com/CAD97/pointer-utils/tree/master/crates%2Frc-borrow%2Fsrc%2Flib.rs). It's currently implemented as the rather opaque `let this = ManuallyDrop::new(Arc::from_raw(ptr)); Arc::clone(&this)`. This allows me to implement it as `Arc::increment_strong_count(ptr); Arc::from_raw(ptr)`, which I think is clearer. It also allows me to avoid the spec lawyer interpretation of `from_raw` "taking ownership" of a strong count and my never undoing that with the current impl; it would probably be more correctly written as (the sadly more error prone (due to holding the `Arc` by-value temporarily)) `let this = Arc::from_raw(ptr); let clone = Arc::clone(&this); Arc::into_raw(this); clone`, such that the strong count is "freed" back into the pool for raw pointers.

TL;DR I'm a 👍 for stabilizing at least the increment functionality, and have another real-world potential usage. Increment has potential pitfalls in implementation.

The same crate contains a use for the weak count version of the same, as well.

(Though I might've just done `clone_raw(*const T) -> Arc<T>`, as the domain there is more clearly on the pointer, and `increment_strong_count(_) = into_raw(clone_raw(_))`, `decrement_strong_count(_) = drop(from_raw(_))`. However, providing the same functionality for weak count is hard to do in that scheme -- `Weak::clone_raw` suggests you need to have a raw weak to clone -- so that's a strong argument for the naming given here.)

One other note: the exposure of this functionality sort of confirms that for the purposes of converting between `*const T` and `Arc<T>`, all `*const T` (derived from an originating `Arc::into_raw`) are equal, and they don't maintain ownership of a specific strong reference they're associated to. (I.e. `a = Arc::new(_); b = Arc::into_raw(a.clone()); c = Arc::into_raw(a); Arc::from_raw(b); Arc::from_raw(b);` is fine.) I think this was intentionally implied by current documentation (and don't see any reasonable argument why it shouldn't be this way), but these functions require this behavior, and help to cement the model.

-----

One doc clarification request: make it clearer that `decrease_strong_count` will drop the innards if it's the last strong count being freed, before digging into the Safety heading. (It could be intuitively assumed to just touch the strong count and not handle the final drop; another argument for spelling it `drop(from_raw)`.) Also,

> but **should not** be called after the final `Arc` has been released [emphasis original]

should probably be "**must not**" to avoid confusion.

**Comment by SimonSapin** [comment]
Sorry I’m so late in the process but I feel this design is flawed.

Reasoning in terms of counter manipulation can lead to off-by-one errors, I think it’s better (especially with Rust’s move semantics) to consider a strong reference as a resource whose ownership can be transferred (or even borrowed, with `&Arc<T>`). Like `Box::from_raw`, `Arc::from_raw` takes ownership away from its input pointer such that it shouldn’t be used anymore.

The motivation discussed in https://github.com/rust-lang/rust/pull/68700#discussion_r396169064 is that unlike `Box`, with `Arc` we can and in some cases want to keep the pointer valid and make a new strong reference out of it instead of "consuming" it. Yes using `Arc::clone` + `mem::forget` is a clunky and confusing way to do this, but I feel that `Arc::increment_strong_count` is also unsafely patching things up at a lower level than necessary.

I don’t see a legitimate use of `increment_strong_count` that is not paired together with `Arc::from_raw`. Is there one? If not I feel that we’d be better off *not* exposing direct counter manipulation and instead add a variation of `from_raw` with the semantics desired here in terms of strong reference ownership.

```rust
impl<T: ?Sized> Arc<T> {
    pub unsafe fn clone_from_raw(ptr: *const T) -> Arc<T> {…}
}
```

(As to `decrement_strong_count`, it looks like it was mostly added for symmetry?)

This might not work, but:

@rfcbot concern clone_from_raw instead

**Comment by Mark-Simulacrum** [comment]
The RawWaker API would need to forget the resulting Arc rather than making use of it in the clone "method". I think most use cases for this API aren't actually creating Arcs directly. The uses I've had historically have also done similar things to the RawWaker API, for example across the FFI boundary where either side wants to increment the counter on a shared resource, due to cloning, but isn't actually going to pass back the Arc (because nothing is going to interact with the Arc directly). That said I don't disagree that the clone_from_raw API is in some cases a bit cleaner; I think I would rather add both though. These methods are largely aliases for 1-3 lines of code, we can expose more than one incrementing method if we want. But I think there are valid uses for both just incrementing and for incrementing and getting a reference.

**Comment by m-ou-se** [comment]
@SimonSapin I think you have to be on rfcbot's list to add concerns.

@rfcbot concern clone_from_raw instead

https://github.com/rust-lang/rust/pull/79285#issuecomment-751333368

**Comment by m-ou-se** [comment]
@SimonSapin Since this is blocked on your concern, can you respond to Mark's comment? Thanks!

**Comment by SimonSapin** [comment]
> The RawWaker API would need to forget the resulting Arc rather than making use of it in the clone "method". I think most use cases for this API aren't actually creating Arcs directly.

Ok, I didn’t realize that.

This concern can be marked as resolved.

**Comment by m-ou-se** [comment]
@rfcbot resolve clone_from_raw instead

File: library/alloc/src/sync.rs [hunk 1]
```diff
@@ -887,8 +885,8 @@ impl<T: ?Sized> Arc<T> {
     /// }
     /// ```
     #[inline]
-    #[unstable(feature = "arc_mutate_strong_count", issue = "71983")]
-    pub unsafe fn incr_strong_count(ptr: *const T) {
+    #[stable(feature = "arc_mutate_strong_count", since = "1.50.0")]
```
Comment (m-ou-se) [inline library/alloc/src/sync.rs:888]:
```suggestion
    #[stable(feature = "arc_mutate_strong_count", since = "1.51.0")]
```

File: library/alloc/src/sync.rs [hunk 2]
```diff
@@ -909,27 +907,25 @@ impl<T: ?Sized> Arc<T> {
...
     ///     assert_eq!(2, Arc::strong_count(&five));
-    ///     Arc::decr_strong_count(ptr);
+    ///     Arc::decrement_strong_count(ptr);
     ///     assert_eq!(1, Arc::strong_count(&five));
     /// }
     /// ```
     #[inline]
-    #[unstable(feature = "arc_mutate_strong_count", issue = "71983")]
-    pub unsafe fn decr_strong_count(ptr: *const T) {
+    #[stable(feature = "arc_mutate_strong_count", since = "1.50.0")]
```
Comment (m-ou-se) [inline library/alloc/src/sync.rs:927]:
```suggestion
    #[stable(feature = "arc_mutate_strong_count", since = "1.51.0")]
```
